go/libraries/doltcore/schema/index_coll.go

```diff
@@ -0,0 +1,241 @@
+	ContainsColumnCollection(cols ...string) bool
+	// ContainsColumnTagCollection returns whether the collection contains an index that has this exact collection and ordering of columns.
+	ContainsColumnTagCollection(tags ...uint64) bool
+	// Count returns the number of indexes in this collection.
+	Count() int
+	// Get returns the index with the given name, or nil if it does not exist.
+	Get(indexName string) Index
+	// DropIndex removes an index from the table metadata.
+	DropIndex(indexName string) (Index, error)
+	// HasIndexes returns whether this collection has any indexes.
+	HasIndexes() bool
+	// RenameIndex renames an index in the table metadata.
+	RenameIndex(oldName, newName string) (Index, error)
+}
+
+type indexCollectionImpl struct {
+	indexes       map[string]*indexImpl
+	colTagToIndex map[uint64][]*indexImpl
+	colToTag      map[string]uint64
+}
+
+func NewIndexCollection(cols *ColCollection) IndexCollection {
+	ixc := &indexCollectionImpl{
+		indexes:       make(map[string]*indexImpl),
+		colTagToIndex: make(map[uint64][]*indexImpl),
+		colToTag:      make(map[string]uint64),
+	}
+	if cols != nil {
+		for tag, col := range cols.TagToCol {
+			ixc.colToTag[col.Name] = tag
+			ixc.colTagToIndex[tag] = nil
+		}
+	}
+	return ixc
+}
+
+func (ixc *indexCollectionImpl) AddIndexCol(indexName string, cols []string, pks []string) (Index, error) {
+	if ixc.Contains(indexName) {
+		return nil, fmt.Errorf("`%s` already exists as an index for this table", indexName)
+	}
+	if ixc.ContainsColumnCollection(cols...) {
+		return nil, fmt.Errorf("cannot create a duplicate index on this table")
+	}
+	index := &indexImpl{
+		name: indexName,
+	}
+	ixc.indexes[indexName] = index
+	tags := make([]uint64, len(cols))
+	for i, col := range cols {
+		tag := ixc.colToTag[col]
+		tags[i] = tag
+		ixc.colTagToIndex[tag] = append(ixc.colTagToIndex[tag], index)
+	}
+	index.tags = tags
+	pkTags := make([]uint64, len(pks))
+	for i, pk := range pks {
+		tag := ixc.colToTag[pk]
+		pkTags[i] = tag
+	}
+	index.pks = pkTags
+	index.allTags = combineAllTags(tags, pkTags)
+	return index, nil
+}
+
+func (ixc *indexCollectionImpl) AddIndexTag(indexName string, tags []uint64, pks []uint64) (Index, error) {
+	if ixc.Contains(indexName) {
+		return nil, fmt.Errorf("`%s` already exists as an index for this table", indexName)
+	}
+	if ixc.ContainsColumnTagCollection(tags...) {
+		return nil, fmt.Errorf("cannot create a duplicate index on this table")
+	}
+	index := &indexImpl{
+		name: indexName,
+		tags: tags,
+		allTags: combineAllTags(tags, pks),
+		pks:  pks,
+	}
+	ixc.indexes[indexName] = index
+	for _, tag := range tags {
+		ixc.colTagToIndex[tag] = append(ixc.colTagToIndex[tag], index)
+	}
+	return index, nil
+}
+
+func (ixc *indexCollectionImpl) AllIndexes() []Index {
+	indexes := make([]Index, len(ixc.indexes))
+	i := 0
+	for _, index := range ixc.indexes {
```

Comment:
Can set i in the for loop expression

Reply:
This wouldn't be `i`, it would be the key. `ixc.indexes` is a map.